### PR TITLE
User Story: Save linearized graph

### LINKED ISSUE #78: Save linearised graph
As a user I want the fastest possible execution time. For this the data flow should be linearised at the first execution and then stored in the api.

### Acceptance Criteria

- The api should store a linearised version of the data flow
- The executor should linearise the data flow only after a change
- The executor should use the linearised version if it exists in the api

### PR DESCRIPTION
**Pull request by silvan-flum**
## Related Issues

- Closes #78 

## Definition of Done

- [x] User Story (All Acceptance Criteria met):
  - [x] The api should store a linearised version of the data flow
  - [x] The executor should linearise the data flow only after a change
  - [x] The executor should use the linearised version if it exists in the api
- [x] Code
  - [x] No more code needed
  - [x] No known bugs
- [x] Clean Code
  - [x] Comments have been created for new code
  - [x] ESLinter found no issues (editor & executor only)
  - [x] SonarLint found no issues (API)
  - [x] No unavoidable Code Smells
- [x] Testing
  - [x] Existing Tests pass
  - [x] New Unit Tests created
  - [x] New Unit Tests pass
- [x] SonarCloud Quality Gate passed

## Definition of more than Done (optional)

- [x] Testing
  - [x] New Integration Tests created
  - [x] New Integration Tests passed

### CODE REVIEW

**Review by DuplosFidibuss — CHANGES_REQUESTED** [review]
Looks good overall. Would be good if another executor expert could have a look at this, too. @S4nvers  could you do that?

File: src/api/Flooq.IntegrationTest/LinearizedGraphTest.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  [TestMethod]
+  public async Task CanPostLinearizedGraph()
+  {
+    var graph = new LinearizedGraph
+    {
+      Id = Guid.NewGuid(),
+      Graph = "Test Graph #1"
+    };
```
[the remark this answers is not in the document]
Comment (silvan-flum) [reply src/api/Flooq.IntegrationTest/LinearizedGraphTest.cs:60]: No, it's different here. The associated graph has the same id as the data flow. Consequently, the id has to be set in a POST operation.

File: src/api/Flooq.Test/Controllers/LinearizedGraphControllerTest.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    Assert.AreSame(actionResult, actionResultReceived);
+  }
+  
+  [TestMethod]
+  public async Task CanGetDataFlow()
+  {
+    _graphServiceMock.Setup(service => service.GetGraph(_graph.Id)).ReturnsAsync(_graph);
+    var graphController = new LinearizedGraphController(_graphServiceMock.Object);
+
+    var actionResultReceived = await graphController.GetGraph(_graph.Id);
```
[the remark this answers is not in the document]
Comment (silvan-flum) [reply src/api/Flooq.Test/Controllers/LinearizedGraphControllerTest.cs:78]: Copy-paste is a skill that has to be learnt...  Changed it.

File: src/api/Flooq.Test/Controllers/LinearizedGraphControllerTest.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    
+    var graph = actionResult.Value;
+    Assert.IsNotNull(graph);
+    Assert.AreSame(_graph, graph);
+
+    actionResult = await graphController.GetGraph(newId);
+    Assert.IsInstanceOfType(actionResult, typeof(ActionResult<LinearizedGraph>));
+    
+    graph = actionResult.Value;
+    Assert.IsNull(graph);
```
[the remark this answers is not in the document]
Comment (silvan-flum) [reply src/api/Flooq.Test/Controllers/LinearizedGraphControllerTest.cs:102]: Done.

File: src/api/Flooq.Test/Services/LinearizedGraphServiceTest.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    Id = Guid.NewGuid(),
+    Graph = ""
+  };
+  
+  [TestInitialize]
+  public async Task TestInitialize()
+  {
+    foreach (var graph in _context.Graphs) _context.Graphs.Remove(graph);
+    await _context.SaveChangesAsync();
+  }
```
[the remark this answers is not in the document]
Comment (silvan-flum) [reply src/api/Flooq.Test/Services/LinearizedGraphServiceTest.cs:29]: Done.

**Review by DuplosFidibuss — APPROVED** [review]

File: src/executor/source/executor/Linearization.ts [hunk 5]
```diff
@@ -1,11 +1,11 @@
-import { Dataflow, Edge, LinearizedDataflow, Node } from '../Dataflow'
+import { Dataflow, Edge, Node } from '../Dataflow'
 
 /**
  * @param dataflow to order in such a way that a iterative execution of the nodes
  * is possible where all inputs are calculated at the time of the execution
  * @returns the linearized dataflow
  */
-export function linearize( dataflow: Dataflow ): LinearizedDataflow {
+export async function linearize( dataflow: Dataflow ): Promise<any> {
```
Comment (S4nvers) [inline src/executor/source/executor/Linearization.ts:8]: - Why is this function async? There is no await or similar in its body.
- Is the any-type in the Promise really necessary or can this be typed even stronger? The function should return a LinearizedDataflow-object only right?

File: src/executor/spec/executor/Linearization.spec.ts [hunk 6]
```diff
@@ -17,13 +17,13 @@ test( 'Linearization of linear graph', () => {
     nodes: testNodes,
   }
 
-  const linearizedFlow = linearize( testDataflow )
+  const linearizedFlow = await linearize( testDataflow )
 
   expect( isLinearized( linearizedFlow ) ).toBe( true )
 
 } )
```
Comment (S4nvers) [inline src/executor/spec/executor/Linearization.spec.ts:24]: Same thing as above. I don't think the linearize-function needs to be async.
Multiple lines in this file are affected, but I don't think every line needs it's own comment

File: src/executor/spec/routes/DataflowRouter.spec.ts [hunk 7]
```diff
@@ -19,8 +25,21 @@ const mockAPI = jest.spyOn( apiInterface, 'getDataflow' )
...
+const mockAPIPostLinearized = jest.spyOn( apiInterface, 'postLinearizedDataflow' )
+
 const mockExecutor = jest.spyOn( Executor, 'execute' )
-  .mockImplementation( ( df, i ) => Promise.resolve( { df, i } ) )
+  .mockImplementation( ( i, ldf ) => Promise.resolve( { i, ldf } ) )
+
+const mockLinearization = jest.spyOn( Linearization, 'linearize' )
+  .mockImplementation( ( dataflow ) => {
+    return Promise.resolve( defaultLinearizedDataflow )
+  } )
```
Comment (S4nvers) [inline src/executor/spec/routes/DataflowRouter.spec.ts:42]: The async linearize-function again

**Review by S4nvers — CHANGES_REQUESTED** [review]

**Review by S4nvers — APPROVED** [review]

**Pull request merged**